src/main/src/main/java/org/vfny/geoserver/global/GeoServerFeatureStore.java

```diff
@@ -96,29 +127,80 @@ public Transaction getTransaction() {
 
     @Override
     public void modifyFeatures(String name, Object attributeValue, Filter filter) throws IOException {
+
+        enforceFeatureRestrictions(name, attributeValue);
+
         filter = makeDefinitionFilter(filter);
 
         store().modifyFeatures(name, attributeValue, filter);
     }
 
     @Override
     public void modifyFeatures(String[] names, Object[] attributeValues, Filter filter) throws IOException {
+
+        enforceFeatureRestrictions(names, attributeValues);
+
         filter = makeDefinitionFilter(filter);
 
         store().modifyFeatures(names, attributeValues, filter);
     }
 
     @Override
     public void modifyFeatures(Name[] attributeNames, Object[] attributeValues, Filter filter) throws IOException {
+
+        String[] names = Stream.of(attributeNames).map(Name::getLocalPart).toArray(String[]::new);
+        enforceFeatureRestrictions(names, attributeValues);
+
         filter = makeDefinitionFilter(filter);
 
         store().modifyFeatures(attributeNames, attributeValues, filter);
     }
 
     @Override
     public void modifyFeatures(Name attributeName, Object attributeValue, Filter filter) throws IOException {
+
+        enforceFeatureRestrictions(attributeName.getLocalPart(), attributeValue);
         filter = makeDefinitionFilter(filter);
 
         store().modifyFeatures(attributeName, attributeValue, filter);
     }
+
+    private void enforceFeatureRestrictions(String name, Object attributeValue) {
+        enforceFeatureRestrictions(new String[] {name}, new Object[] {attributeValue});
+    }
+
+    private void enforceFeatureRestrictions(String[] names, Object[] attributeValues) {
+        Map<String, Object> updates = new HashMap<>();
+        for (int i = 0; i < names.length; i++) {
+            updates.put(names[i], attributeValues[i]);
+        }
+
+        for (PropertyDescriptor propertyDescriptor : schema.getDescriptors()) {
+            Object value = updates.get(propertyDescriptor.getName().toString());
+            Class<?> binding = propertyDescriptor.getType().getBinding();
+
+            Object convertedValue = Converters.convert(value, binding);
+
+            if (isValueViolatingRestrictions(
+                    convertedValue, propertyDescriptor.getType().getRestrictions())) {
+                throw new IllegalArgumentException(String.format(
```

Comment:
As above, ``Types.validate( convertedValue, propertyDescriptor )`` should offer to check restrictions from super types and so on.